implement plugins for custom devices

ardera opened the pull request:
Implements plugin for custom devices: See https://github.com/flutter/flutter/issues/86864
See also the design doc: https://flutter.dev/go/plugin-support-for-custom-embeddings

### Summary of the changes

`custom_devices/custom_device_config.dart`:
- add 3 new keys:
  - `embedder`, the name of the embedder
  - `configureNativeProject` (should basically invoke `cmake` in the native project directory)
  - `buildNativeProject` (should basically invoke `ninja` in the native project directory)
  - native project directory is the `x-flutter-pi` subdir of the flutter project directory
  - (if the embedder is called flutter-pi)

`custom_devices/custom_device.dart`:
- add methods for invoking the `configureNativeProject` and `buildNativeProject` commands of the custom device config
- check the return value of the postBuildCommand, return LaunchResult.failed() if not successful
- check the return value of the installCommand, return LaunchResult.failed() if not successful

`flutter_plugins.dart`:
- expand writing of .flutter-plugins-dependencies for custom devices plugins
- expand the generated dart plugin registrant for custom devices plugins
  - for example, the `flutter-pi` variant of a plugin is activated if `-Dflutter.customPlatform=flutter-pi` was given as a dart define, or the `FLUTTER_CUSTOM_PLATFORM` environment variable equals `flutter-pi` (needed for debug mode)
- also creating plugin symlinks for custom devices plugins
- allow resolving platform interface resolutions for custom devices plugins
- add method to find all plugins for custom embedder with a given platform key

`project.dart`:
- expand `FlutterProjectFactory` and `FlutterProject` for custom devices support
- `FlutterProject.customEmbedderProjects` lists all custom embedder subprojects of this flutter project
- to do that, it needs to know which custom embedders are currently configured
- it gets that info from the flutter project factory (which just queries the CustomDevicesConfig)    

Tests:
- add tests for all the changes
- add makeProjectFactory inside test/src/common.dart with creates a project factory with some default arguments, make all tests use that instead of constructing a FlutterProjectFactory themselves

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ardera committed:
implement plugins for custom devices

custom_devices/custom_device_config.dart:
- add 3 new keys:
  - `embedder`, the name of the embedder
  - `configureNativeProject` (should basically invoke `cmake` in the
  native project directory)
  - `buildNativeProject` (should basically invoke `ninja` in the
  native project directory)
  - native project directory is the `x-flutter-pi` subdir of the
  flutter project directory
  - (if the embedder is called flutter-pi)
- add `-tt` ssh argument on linux and macOS to force a pseudo-terminal
(so remote will get a SIGHUP when the host kills ssh, otherwise the
embedder won't terminate)
- add trailing comma to argument lists

custom_devices/custom_device.dart:
- add methods for invoking the `configureNativeProject` and
`buildNativeProject` commands of the custom device config
- check the return value of the postBuildCommand, return
LaunchResult.failed() if not successful
- check the return value of the installCommand, return
LaunchResult.failed() if not successful

plugins.dart:
- add validation, parsing for pubspec.yaml custom device stuff

flutter_plugins.dart:
- expand writing of .flutter-plugins-dependencies for custom devices
plugins
- expand the generated dart plugin registrant for custom devices
plugins
  - for example, the `flutter-pi` variant of a plugin is activated if
  `-Dflutter.customPlatform=flutter-pi` was given as a dart define,
  or the `FLUTTER_CUSTOM_PLATFORM` environment variable equals
  `flutter-pi` (needed for debug mode)
- also creating plugin symlinks for custom devices plugins
- allow resolving platform interface resolutions for custom devices
plugins
- add method to find all plugins for custom embedder with a given
platform key

project.dart:
- expand `FlutterProjectFactory` and `FlutterProject` for custom
devices support
- `FlutterProject.customEmbedderProjects` lists all custom embedder
subprojects of this flutter project
- to do that, it needs to know which custom embedders are currently
configured
- it gets that info from the flutter project factory (which just
queries the CustomDevicesConfig)

- add tests for all the changes
- add makeProjectFactory inside test/src/common.dart with creates a
project factory with some default arguments, make all tests use that
instead of constructing a FlutterProjectFactory themselves

packages/flutter_tools/lib/src/custom_devices/custom_device.dart
packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart
packages/flutter_tools/test/commands.shard/hermetic/custom_devices_test.dart
packages/flutter_tools/test/general.shard/android/android_device_test.dart
packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart
packages/flutter_tools/test/general.shard/crash_reporting_test.dart
packages/flutter_tools/test/general.shard/custom_devices/custom_device_test.dart
packages/flutter_tools/test/general.shard/custom_devices/custom_devices_config_test.dart
packages/flutter_tools/test/general.shard/custom_devices/custom_devices_plugins_test.dart
packages/flutter_tools/test/general.shard/github_template_test.dart
packages/flutter_tools/test/general.shard/linux/linux_device_test.dart
packages/flutter_tools/test/general.shard/macos/macos_device_test.dart
packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
packages/flutter_tools/test/general.shard/windows/windows_device_test.dart

ardera commented:
(ping) @christopherfujino

christopherfujino commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -53,9 +56,8 @@ class CustomDeviceRevivalException implements Exception {
-    String fieldDescription,
-    String expectedValueDescription
-  ) : message = 'Expected $fieldDescription to be $expectedValueDescription.';
+      String fieldDescription, String expectedValueDescription)
```
prefer previous formatting, per our style guide https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

ardera replied:
reverted the formatting changes

christopherfujino commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -94,117 +96,91 @@ class CustomDeviceConfig {
+        typedMap[_kForwardPortCommand],
+        _kForwardPortCommand,
+        'null or array of strings with at least one element',
+        minLength: 1);
```
can you update this and the other argument lists below split onto multilines to have a trailing comma after the last argument and then a newline before the close paren, per our style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

ardera replied:
ok, did that (and reverted the formatting changes)

christopherfujino commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -18,13 +18,13 @@ class Plugin {
-       assert(pluginDartClassPlatforms != null),
-       assert(dependencies != null),
-       assert(isDirectDependency != null);
+  })  : assert(name != null),
```
can you revert the formatting only diffs to this file? They are not all in line with our [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting) and make this harder to review

ardera replied:
reverted them

christopherfujino commented on packages/flutter_tools/lib/src/platform_plugins.dart:
```diff
@@ -462,6 +462,65 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
+    required String pluginPath,
+    required FileSystem fileSystem
+  }) {
+    assert(validate(yaml));
```
so asserts don't run in the tool unless you explicitly enable them by editing the shell script entrypoint. should we run this validation always?

ardera replied:
The check in https://github.com/flutter/flutter/blob/90fa1b3be0e678704b709d1dd85cba0723814faf/packages/flutter_tools/lib/src/plugins.dart#L280 should've already made sure the yaml is valid (except I forgot to add some validation there, but I did that now)

christopherfujino commented:
@stuartmorgan can you review the changes to:

- packages/flutter_tools/lib/src/flutter_plugins.dart
- packages/flutter_tools/lib/src/platform_plugins.dart

ardera committed:
fixup after rebasing
platform_plugins.dart:
- remove `CustomEmbedderPlugin.isNative`, no longer present in parent
- support `hasDart`, `hasFfi`, `hasMethodChannel` (just add some boolean flags in the pubspec for the latter two)

flutter_plugins.dart:
- rework findNativeCustomEmbedderPlugins

ardera commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -5,6 +5,7 @@
 // @dart = 2.8
 
 import 'dart:async';
+import 'dart:io';
```
fixed it

ardera commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -773,18 +846,64 @@ class CustomDevice extends Device {
+        final CustomEmbedderProject project = FlutterProject.current().customEmbedderProjects.singleWhere((CustomEmbedderProject p) => p.embedderName == _config.embedderName);
+
+        if (!project.baseDirectory.existsSync()) {
+          // TODO(ardera): Add link here
```
added the [link](https://github.com/flutter/flutter/wiki/Using-custom-embedders-with-the-Flutter-CLI). Though it doesn't contain documentation for plugin support right now, but I opened https://github.com/flutter/flutter/issues/102133 to track that.

ardera commented:
Sorry for the delay!

(had to squash because I had 1000s of merge conflicts because of formatting changes in my commits that I reverted in later commits, sry if that messes up the review or something)

ardera committed:
fixes
- fix passing of removed parameter `fileSystem` for `CustomEmbedderPlugin`
- remove unused import in test

ardera committed:
fix test
- documentation URL was updated to point to an actual website, use that in the test too

ardera committed:
update tests
- custom device config template has changed, update test to match against the updated template

stuartmorgan commented on packages/flutter_tools/lib/src/platform_plugins.dart:
```diff
@@ -532,6 +535,78 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
   }
 }
 
+class CustomEmbedderPlugin extends PluginPlatform implements NativeOrDartPlugin {
```
Related to my comments on the design document: I don't think that this should implement `NativeOrDartPlugin`, because it's a specific set of implementation keys, and keys are 100% platform-dependent. An alternate web embedder, for instance, would definitely not be a `NativeOrDartPlugin`.

stuartmorgan commented on packages/flutter_tools/lib/src/platform_plugins.dart:
```diff
@@ -532,6 +535,78 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
+    required this.embedderName,
+    required this.name,
+    required this.pluginPath,
+    this.dartPluginClass,
```
Per the discussion in the doc, there's currently no approach to actually implement this, so it shouldn't be a universal key.

stuartmorgan commented on packages/flutter_tools/lib/src/platform_plugins.dart:
```diff
@@ -532,6 +535,78 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
+    required this.pluginPath,
+    this.dartPluginClass,
+    this.defaultPackage,
+    bool? ffiPlugin,
```
This may not apply to all embedders.

stuartmorgan commented on packages/flutter_tools/lib/src/platform_plugins.dart:
```diff
@@ -532,6 +535,78 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
+    this.dartPluginClass,
+    this.defaultPackage,
+    bool? ffiPlugin,
+    bool? methodChannelPlugin
```
Nor this.

I'm not even clear on what this is; it's unused AFAICT.

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -235,14 +257,22 @@ class Plugin {
+    List<String> platforms,
+    String pluginClass,
+    String androidPackage, [
+    List<String> emptyPlatforms = const <String>[],
```
This needs explanation in the method comment.

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -320,6 +350,11 @@ class Plugin {
     }
+    yaml.forEach((dynamic key, dynamic value) {
+      if (key is String && key.startsWith('x-')) {
+        isInvalid(key, CustomEmbedderPlugin.validate);
```
I'm not sure we can meaningfully validate custom plugin sections; they could contain anything.

ardera commented:
I apologize for the delay. I hope I'll get to it this week or next week.

eximius313 commented:
@ardera do you have plans regarding this feature?

christopherfujino commented:
Hey @ardera, I am going to close this PR just to get it off our PR review queue. Feel free to re-open if/when you have time to pick it up again.
